python/heterocl/schedule.py

```diff
@@ -292,66 +291,327 @@ def to(self, tensors, dst, src=None, axis=0,
+
+        # inter-stage data movement
+        if not (isinstance(dst, Device) or isinstance(dst, DevMediaPair)):
+            # 1. handle inter-HCL-module data streaming 
+            if isinstance(src.op, _tensor.PlaceholderOp) and isinstance(dst.op, _tensor.PlaceholderOp): 
+                # search the kernel calls globally and find the target tensor
+                print("[ INFO ] inter-kernel streaming for target tensor")
+                src_stage_name = src.op.name.split(".")[1]
+                dst_stage_name = dst.op.name.split(".")[1]
+                if src_stage_name not in self.mod_calls:
+                    raise HCLError("{} is not called".format(src_stage_name))
+                if dst_stage_name not in self.mod_calls:
+                    raise HCLError("{} is not called".format(dst_stage_name))
+                if len(self.mod_calls[src_stage_name]) > 1:
+                    raise HCLError("{} has more than one call sites".format(src_stage_name))
+                if len(self.mod_calls[dst_stage_name]) > 1:
+                    raise HCLError("{} has more than one call sites".format(dst_stage_name))
+                    
+                def get_overlap(a, b):
+                    return list(set(a) & set(b))[0]
+
+                tensor = get_overlap(self.mod_calls[dst_stage_name][0],
+                    self.mod_calls[src_stage_name][0])
+                tensor = tensor.tensor
+
+            # 2. check whether the streaming channel has been created 
+            # from the target tensor to the destination stage
+            dst_stages = set()
+            if tensor.name in self.stream_channels.keys():
+                dst_stages = self.stream_channels[tensor.name]
+            size = len(dst_stages)
+            t = (src.op.name, dst.op.name)
+            dst_stages.add(t)
+            if size == len(dst_stages):
+                print("[ Warning ] " + 
+                    "the tensor {} has been streamed to stage {}... Ignored"
+                    .format(tensor.name, dst.op.name))
+                return
+            self.stream_channels[tensor.name] = dst_stages
+
+        # save tensor and source stage to support .to chain
+        self.cascade_tensor = tensor
+        if not move_to_device: self.hold_source_stage = dst
+        else: self.hold_source_stage = None
+
+        # target can be stage or tensor
+        # the pre-processing is finished here
+        # run the TVM FFI APIs to annotate the DFG with different modes
+        # --------------------------------------------
+        ret = None
+        # 1. Place a target tensor to device
+        if (isinstance(dst, Device) or isinstance(dst, DevMediaPair)):
+            is_pair = False if isinstance(dst, Device) else True
+            media = dst.media if is_pair else dst.DRAM.media
+
+            # Check the device ID (for multi-device platform) 
+            dev_id = dst.dev.get_dev_id() if is_pair else dst.get_dev_id()
```

Comment:
What's a multi-device platform? Do we support that right now?

Reply:
No. will remove it